File: erpnext_custom/erpnext_custom/report/report_for_manager/report_for_manager.py
import frappe


def execute(filters=None):
    if not filters or not filters.get("doctype_name"):
        return [], []

    doctype = filters.get("doctype_name")

    roles = frappe.get_roles(frappe.session.user)
    allowed_doctypes = []

    if "Administrator" in roles or "Account Manager" in roles:
        allowed_doctypes = ["Sales Order", "Sales Invoice", "Purchase Order", "Purchase Invoice"]
    elif roles == "Sales Manager":
        allowed_doctypes = ["Sales Order", "Sales Invoice"]
    elif roles == "Purchase Manager":
        allowed_doctypes = ["Purchase Order", "Purchase Invoice"]
    elif roles == "Accounts Manager":
        allowed_doctypes = ["Sales Invoice", "Purchase Invoice"]
         
    if doctype not in allowed_doctypes:
        frappe.throw(f"You do not have permission to view {doctype}")

    doc_filters = {}
    if filters.get("workflow_state"):
        doc_filters["workflow_state"] = filters.get("workflow_state")

    columns = [
        {"label": "Select", "fieldname": "select_row", "fieldtype": "HTML"},
        {"label": "Document", "fieldname": "name", "fieldtype": "Link", "options": doctype},
        {"label": "Party", "fieldname": "party", "fieldtype": "Data"},
        {"label": "Grand Total", "fieldname": "grand_total", "fieldtype": "Currency"},
        {"label": "Workflow State", "fieldname": "workflow_state", "fieldtype": "Data"},
    ]

    # --- Fetch Data ---
    if doctype == "Sales Invoice":
        data = frappe.get_all(
            "Sales Invoice",
            fields=["name", "customer as party", "grand_total", "workflow_state"],
            filters=doc_filters
        )

    elif doctype == "Sales Order":
        data = frappe.get_all(
            "Sales Order",
            fields=["name", "customer as party", "grand_total", "workflow_state"],
            filters=doc_filters
        )

    elif doctype == "Purchase Invoice":
        data = frappe.get_all(
            "Purchase Invoice",
            fields=["name", "supplier as party", "grand_total", "workflow_state"],
            filters=doc_filters
        )

    elif doctype == "Purchase Order":
        data = frappe.get_all(
            "Purchase Order",
            fields=["name", "supplier as party", "grand_total", "workflow_state"],
            filters=doc_filters
        )

    # --- Add checkbox for each row ---
    for row in data:
        row["select_row"] = f'<input type="checkbox" class="row-select" data-name="{row["name"]}">'

    return columns, data

import frappe

@frappe.whitelist()
def get_workflow_states(doctype_name):
    workflows = frappe.get_all(
        "Workflow", 
        filters={"document_type": doctype_name}, 
        pluck="name"
    )
    if workflows:
        states = frappe.get_all(
            "Workflow Document State",
            filters={"parent": workflows[0]},
            pluck="state"
        )
    
    states = (set(states))
    return states


@frappe.whitelist()
def update_workflow_state(doctype_name, docs, action):
    """Apply a specific workflow action (chosen by Manager) on multiple docs."""
    import json
    from frappe.model.workflow import apply_workflow

    docs = json.loads(docs) if isinstance(docs, str) else docs
    results = []

    for docname in docs:
        try:
            doc = frappe.get_doc(doctype_name, docname)

            # Apply the chosen action
            apply_workflow(doc, action)
            frappe.db.commit()

            results.append({"doc": docname, "status": f"Moved via {action}"})

        except Exception as e:
            frappe.log_error(frappe.get_traceback(), f"Workflow Update Failed for {docname}")
            results.append({"doc": docname, "status": f"Error: {str(e)}"})

    return results

@frappe.whitelist()
def get_workflow_actions(doctype_name):
    """Return all possible workflow actions for a given Doctype."""
    actions = []
    workflows = frappe.get_all(
        "Workflow",
        filters={"document_type": doctype_name},
        pluck="name"
    )
    if workflows:
        actions = frappe.get_all(
            "Workflow Transition",
            filters={"parent": ["in", workflows]},
            pluck="action"
        )
    return sorted(set(actions))

File: erpnext_custom/erpnext_custom/report/report_for_manager/report_for_manager.js
// frappe.query_reports["Report For Manager"] = {
//     filters: [
//         {
//             fieldname: "doctype_name",
//             label: "Document Type",
//             fieldtype: "Select",
//             options: [" ", "Sales Order", "Sales Invoice", "Purchase Order", "Purchase Invoice"],
//             default: " ",
//             reqd: 1
//         },
//         {
//             fieldname: "workflow_state",
//             label: "Workflow State",
//             fieldtype: "Select",
//             options: [" "],
//             default: " ",
//             reqd: 1
//         }
//     ],

//     onload: function(report) {
//         // 🔹 Bind onchange for Doctype filter
//         report.get_filter('doctype_name').$input.on("change", function() {
//             let doctype = frappe.query_report.get_filter_value("doctype_name");
//             console.log("📢 onchange fired, doctype =", doctype);

//             if (!doctype || doctype === " ") return;

//             frappe.call({
//                 method: "erpnext_custom.erpnext_custom.report.report_for_manager.report_for_manager.get_workflow_states",
//                 args: { doctype_name: doctype },
//                 callback: function(r) {
//                     console.log("📢 API response =", r);

//                     if (r.message) {
//                         let state_filter = frappe.query_report.get_filter('workflow_state');
//                         state_filter.df.options = [" "].concat(r.message);
//                         state_filter.df.default = " ";
//                         state_filter.refresh();
//                         state_filter.set_input(" ");
//                     }
//                 }
//             });
//         });

//         // 🔹 Add Workflow Action button
//         report.page.add_inner_button(__("Workflow Action"), function() {
//             let selected = [];
//             $(".row-select:checked").each(function() {
//                 selected.push($(this).data("name"));
//             });

//             if (selected.length === 0) {
//                 frappe.msgprint("Please select at least one document.");
//                 return;
//             }

//             let doctype = frappe.query_report.get_filter_value("doctype_name");
//             console.log("Selected documents:", doctype, selected);

//             frappe.call({
//                 method: "erpnext_custom.erpnext_custom.report.report_for_manager.report_for_manager.update_workflow_state",
//                 args: {
//                     doctype_name: doctype,
//                     docs: selected
//                 },
//                 callback: function(r) {
//                     if (!r.exc) {
//                         frappe.msgprint("Workflow updated successfully.");
//                         frappe.query_report.refresh();
//                     }
//                 }
//             });
//         });
//     }
// };



frappe.query_reports["Report For Manager"] = {
    filters: [
        {
            fieldname: "doctype_name",
            label: "Document Type",
            fieldtype: "Select",
            options: [" ", "Sales Order", "Sales Invoice", "Purchase Order", "Purchase Invoice"],
            default: " ",
            reqd: 1
        },
        {
            fieldname: "workflow_state",
            label: "Workflow State",
            fieldtype: "Select",
            options: [" "],
            default: " ",
            reqd: 1
        }
    ],

    onload: function(report) {
        // Bind onchange for Document Type filter
        report.get_filter('doctype_name').$input.on("change", function() {
            let doctype = frappe.query_report.get_filter_value("doctype_name");
            if (!doctype || doctype === " ") return;

            // Load workflow states
            frappe.call({
                method: "erpnext_custom.erpnext_custom.report.report_for_manager.report_for_manager.get_workflow_states",
                args: { doctype_name: doctype },
                callback: function(r) {
                    if (r.message) {
                        let state_filter = frappe.query_report.get_filter('workflow_state');
                        state_filter.df.options = [" "].concat(r.message);
                        state_filter.df.default = " ";
                        state_filter.refresh();
                        state_filter.set_input(" ");
                    }
                }
            });

            // Load workflow actions (for dropdown button)
            frappe.call({
                method: "erpnext_custom.erpnext_custom.report.report_for_manager.report_for_manager.get_workflow_actions",
                args: { doctype_name: doctype },
                callback: function(r) {
                    if (r.message) {
                        // remove old button first
                        report.page.clear_inner_toolbar();

                        // create dropdown
                        report.page.add_inner_button(__("Workflow Actions"), null, "Actions");

                        r.message.forEach(action => {
                            report.page.add_inner_button(action, function() {
                                let selected = [];
                                $(".row-select:checked").each(function() {
                                    selected.push($(this).data("name"));
                                });

                                if (selected.length === 0) {
                                    frappe.msgprint("Please select at least one document.");
                                    return;
                                }

                                frappe.call({
                                    method: "erpnext_custom.erpnext_custom.report.report_for_manager.report_for_manager.update_workflow_state",
                                    args: {
                                        doctype_name: doctype,
                                        docs: selected,
                                        action: action   // 👈 send chosen action
                                    },
                                    callback: function(r) {
                                        if (!r.exc) {
                                            frappe.msgprint(`Workflow action '${action}' applied successfully.`);
                                            frappe.query_report.refresh();
                                        }
                                    }
                                });
                            }, "Workflow Actions");
                        });
                    }
                }
            });
        });
    }
};
